### PR TITLE
Kafka Histogram Listener 

### LINKED ISSUE #21414: Add Kafka Histo mode Listener
Issue to add histo mode streaming support with Kafka.

Details to be added.

### PR DESCRIPTION
**Pull request by LamarMoore**
**Description of work.**

Lost confidence in [this PR](https://github.com/mantidproject/mantid/pull/23310) due to merge conflicts. I have cherry-picked the relevant changes and added them here.

**To test:**

Instructions for testing:

- Get Nexus Streamer: https://github.com/ScreamingUdder/isis_nexus_streamer_for_mantid/
- Get Histo Streamer: https://github.com/DMSC-Instrument-Data/kafka_histo_streamer/
- Run Nexus Streamer using `docker-compose up`
- Wait for messages:
```
producer_1   | Publishing new run:
producer_1   | Run number: 1, Instrument name: SANS2D, Start time: 2018-11-19T16:31:23
```
- Run Histo Streamer using `mantidpython run.py`
- In Mantid, run StartLiveData:

```
StartLiveData(FromNow=True, FromTime=False, FromStartOfRun=False, UpdateEvery=5.0,
              Instrument="SANS2D", Listener="KafkaHistoListener",
              Address="172.18.0.3:9092", RunTransitionBehavior="Stop",
              AccumulationMethod="Replace", OutputWorkspace="histo", InstrumentName="SANS2D")
```

Fixes #21414 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/LiveData/inc/MantidLiveData/Kafka/KafkaHistoListener.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+//----------------------------------------------------------------------
+// Forward declarations
+//----------------------------------------------------------------------
+
+namespace Mantid {
+namespace LiveData {
+class KafkaHistoStreamDecoder;
+
+/**
+  Implementation of a live listener to consume messages from the Kafka system
```
Comment (OwenArnold) [inline Framework/LiveData/inc/MantidLiveData/Kafka/KafkaHistoListener.h]: I think this comment is a copy-paste error @LamarMoore @MikeHart85

File: Framework/LiveData/src/Kafka/KafkaHistoListener.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+DECLARE_LISTENER(KafkaHistoListener)
+
+KafkaHistoListener::KafkaHistoListener() {
+  declareProperty("InstrumentName", "");
+}
+
+/// @copydoc ILiveListener::connect
+bool KafkaHistoListener::connect(const Poco::Net::SocketAddress &address) {
+  auto broker = std::make_shared<KafkaBroker>(address.toString());
```
Comment (OwenArnold) [inline Framework/LiveData/src/Kafka/KafkaHistoListener.cpp]: Why is the broker shared? Is this required?

File: Framework/LiveData/src/Kafka/KafkaHistoListener.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+/// @copydoc ILiveListener::isConnected
+bool KafkaHistoListener::isConnected() {
+  return (m_decoder ? m_decoder->isCapturing() : false);
+}
+
+/// @copydoc ILiveListener::runStatus
+API::ILiveListener::RunStatus KafkaHistoListener::runStatus() {
+  if (!m_decoder) {
+    g_log.error("KafkaHistoListener::runStatus(): Kafka is not connected");
+    throw Kernel::Exception::InternetError("Kafka is not connected");
```
Comment (OwenArnold) [inline Framework/LiveData/src/Kafka/KafkaHistoListener.cpp]: Bit weird mixing return flags and exceptions being thrown. But Ok.

File: Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+ * @param broker A reference to a Broker object for creating topic streams
+ * @param histoTopic The name of the topic streaming the histo data
+ * @param spDetTopic The name of the topic streaming the spectrum-detector
+ * run mapping
+ */
+KafkaHistoStreamDecoder::KafkaHistoStreamDecoder(
+    std::shared_ptr<IKafkaBroker> broker, const std::string &histoTopic,
+    const std::string &instrumentName)
+    : m_broker(broker), m_histoTopic(histoTopic),
+      m_instrumentName(instrumentName), m_histoStream(), m_workspace(),
```
Comment (OwenArnold) [inline Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: Is there any check on the histoTopic or instrumentName being valid? Do we need to check that `KafkaHistoStreamDecoder` is not being initialised with junk?

File: Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+  m_thread = std::thread([this]() { this->captureImpl(); });
+  m_thread.detach();
+}
+
+/**
+ * Stop capturing from the stream. This is a blocking call until the capturing
+ * function has completed
+ */
+void KafkaHistoStreamDecoder::stopCapture() {
```
Comment (OwenArnold) [inline Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: As this is called by the destructor it should be marked noexcept

File: Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+
+  if (!m_capturing) {
+    throw Exception::NotYet("Local buffers not initialized.");
+  }
+
+  if (m_buffer.empty()) {
+    throw Exception::NotYet("No message to process yet.");
+  }
+
+  auto histoMsg = GetEventHistogram(m_buffer.c_str());
```
Comment (OwenArnold) [inline Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: Comment about this being a flat buffer struct?

File: Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+
+  auto shape = histoMsg->current_shape();
```
Comment (OwenArnold) [inline Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: minor point, but do these need to be `const auto` since we only read, no modification required.

File: Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+ * Start decoding data from the streams into the internal buffers.
+ * Implementation designed to be entry point for new thread of execution.
+ * It catches all thrown exceptions.
+ */
+void KafkaHistoStreamDecoder::captureImpl() {
+  m_capturing = true;
+  try {
+    captureImplExcept();
+  } catch (std::exception &exc) {
+    m_exception = boost::make_shared<std::runtime_error>(exc.what());
```
Comment (OwenArnold) [inline Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: Can't this be a `std::unique_ptr` or a `boost::optional<std::runtime_error>` can't see what sharing has to do with this.

File: Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+  API::MatrixWorkspace_sptr workspace;
+
+  try {
+    auto alg = API::AlgorithmManager::Instance().createUnmanaged(
+        "LoadEmptyInstrument");
+    // Do not put the workspace in the ADS
+    alg->setChild(true);
+    alg->initialize();
+    alg->setPropertyValue("InstrumentName", m_instrumentName);
+    alg->setPropertyValue("OutputWorkspace", "ws");
```
Comment (OwenArnold) [inline Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: I think "ws" is a dummy value given the `Algorithm` object setup, maybe indicate that for future readers.

File: Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+    alg->execute();
+    workspace = alg->getProperty("OutputWorkspace");
+  } catch (std::exception &exc) {
+    g_log.error() << "Error loading empty instrument '" << m_instrumentName
+                  << "': " << exc.what() << "\n";
+    throw;
+  }
+
+  return boost::dynamic_pointer_cast<DataObjects::Workspace2D>(workspace);
```
Comment (OwenArnold) [inline Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: Is this guranteed by the algorithm but not enforced by the property type for the return workspace? i.e. any chance this will fail?

File: Framework/LiveData/src/Kafka/private/Schema/ai33_det_count_imgs_generated.h [hunk 11]
```diff
@@ -0,0 +1,111 @@
+// automatically generated by the FlatBuffers compiler, do not modify
```
Comment (OwenArnold) [inline Framework/LiveData/src/Kafka/private/Schema/ai33_det_count_imgs_generated.h]: Agreed that this should not be modified, but how does one find the source schema for this type @MikeHart85 ?

**Review by OwenArnold — CHANGES_REQUESTED** [review]
Thanks for the work on this more lately @LamarMoore and prior to that @MikeHart85 

Aside from the comments made. There is no documentation with this. This should definitely form part of the release notes. We must provide good documentation for this so that the DA DMSC group can make sense of how to use it with `StartLiveData`

File: Framework/LiveData/inc/MantidLiveData/Kafka/KafkaHistoListener.h (hunk 1 above)
Comment (MikeHart85) [reply Framework/LiveData/inc/MantidLiveData/Kafka/KafkaHistoListener.h]: Yes this, like much of this code, was copied from:

* [KafkaEventListener.h](https://github.com/mantidproject/mantid/blob/fcc35cce4241f26f627100f008655ebe7036720b/Framework/LiveData/inc/MantidLiveData/Kafka/KafkaEventListener.h)
* [KafkaEventListener.cpp](https://github.com/mantidproject/mantid/blob/fcc35cce4241f26f627100f008655ebe7036720b/Framework/LiveData/src/Kafka/KafkaEventListener.cpp)
* [KafkaEventStreamDecoder.h](https://github.com/mantidproject/mantid/blob/fcc35cce4241f26f627100f008655ebe7036720b/Framework/LiveData/inc/MantidLiveData/Kafka/KafkaEventStreamDecoder.h)
* [KafkaEventStreamDecoder.cpp](https://github.com/mantidproject/mantid/blob/fcc35cce4241f26f627100f008655ebe7036720b/Framework/LiveData/src/Kafka/KafkaEventStreamDecoder.cpp)

This comment should be updated to reflect that this listener works with histograms rather than events.

File: Framework/LiveData/src/Kafka/KafkaHistoListener.cpp (hunk 2 above)
Comment (MikeHart85) [reply Framework/LiveData/src/Kafka/KafkaHistoListener.cpp]: Copy-pasted from KafkaEventListener. 

Looks like it was unique at first, but was changed to shared by @matthew-d-jones in https://github.com/mantidproject/mantid/commit/00714fbbc540fd2b2bd1b70b254a73876a82b19c as part of #18450.

File: Framework/LiveData/src/Kafka/KafkaHistoListener.cpp (hunk 3 above)
Comment (MikeHart85) [reply Framework/LiveData/src/Kafka/KafkaHistoListener.cpp]: Agreed, this is weird. Can't remember adding this, but also don't see it in the event listener.

Might be worthwhile seeing if there's a better solution. The `RunStatus` enum also has `NoRun` and `BeginRun` elements. Maybe returning `NoRun` would be more appropriate?

File: Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp (hunk 4 above)
Comment (MikeHart85) [reply Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: I think what would happen with an invalid instrument name is `createBufferWorkspace` would throw an exception due to failing to load the instrument, which would then be caught in `KafkaHistoListener::connect`.

Not sure how we would detect an invalid histoTopic... what makes it invalid? If it doesn't exist yet, I think Kafka will just create one when we connect to it. We just won't get any data unless a producer connects to it as well.

File: Framework/LiveData/src/Kafka/KafkaHistoListener.cpp (hunk 2 above)
Comment (matthew-d-jones) [reply Framework/LiveData/src/Kafka/KafkaHistoListener.cpp]: I think I may have had problems with gmock when doing dependency injection with `unique_ptr`.

File: Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp (hunk 4 above)
Comment (matthew-d-jones) [reply Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: We've disabled "automatic" topic creation for the production cluster, so `rdkafka` would return an error code when you try to `assign` the non-existent topic.

File: Framework/LiveData/src/Kafka/private/Schema/ai33_det_count_imgs_generated.h (hunk 11 above)
Comment (matthew-d-jones) [reply Framework/LiveData/src/Kafka/private/Schema/ai33_det_count_imgs_generated.h]: Schemas are at https://github.com/ess-dmsc/streaming-data-types/
maybe this is a good place to document that.

File: Framework/LiveData/src/Kafka/KafkaHistoListener.cpp (hunk 2 above)
Comment (OwenArnold) [reply Framework/LiveData/src/Kafka/KafkaHistoListener.cpp]: Where exactly @matthew-d-jones?

File: Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp (hunk 9 above)
Comment (MikeHart85) [reply Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: Yes, and perhaps there's a better way to specify "this name doesn't matter"? Or maybe it's not necessary to provide a name at all? I think I just did this quickly and meant to have a closer look at it later but forgot.

File: Framework/LiveData/src/Kafka/KafkaHistoListener.cpp (hunk 2 above)
Comment (OwenArnold) [reply Framework/LiveData/src/Kafka/KafkaHistoListener.cpp]: Would be worth a retry. I assume you are ending up (in the mock) pointing to invalid memory after the `unique_ptr` is moved. Perhaps avoiding `make_unique` or at least keeping the address of the held pointer is a better way to construct the mock?

File: Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp (hunk 4 above)
Comment (MikeHart85) [reply Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: Is there a way to query whether a topic exists?

Comment (matthew-d-jones) [reply Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: Looks like it is already handled by the subscriber class: https://github.com/mantidproject/mantid/blob/1dc90bf3242ef7201883e96adf85eca2323ffb72/Framework/LiveData/src/Kafka/KafkaTopicSubscriber.cpp#L310

File: Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp (hunk 9 above)
Comment (LamarMoore) [reply Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: This is a mandatory algorithm property. I think you have to supply this for the algorithm to work.

**Comment by OwenArnold** [comment]
@LamarMoore @MikeHart85. This is lacking tests instructions. Despite being able to listen to kafka events, I would need to be able to generate/stream histograms over histograms, for which I think @MikeHart85 has a script?

Also I note that there is no user documentation update as part of this.

Other than that, the code looks to be in a good state.

**Comment by MikeHart85** [comment]
Instructions for testing:

- Get Nexus Streamer: https://github.com/ScreamingUdder/isis_nexus_streamer_for_mantid/
- Get Histo Streamer: https://github.com/DMSC-Instrument-Data/kafka_histo_streamer/
- Run Nexus Streamer using `docker-compose up`
- Wait for messages:
```
producer_1   | Publishing new run:
producer_1   | Run number: 1, Instrument name: SANS2D, Start time: 2018-11-19T16:31:23
```
- Run Histo Streamer using `mantidpython run.py`
- In Mantid, run StartLiveData:

```
StartLiveData(FromNow=True, FromTime=False, FromStartOfRun=False, UpdateEvery=5.0,
              Instrument="SANS2D", Listener="KafkaHistoListener",
              Address="172.18.0.3:9092", RunTransitionBehavior="Stop",
              AccumulationMethod="Replace", OutputWorkspace="histo", InstrumentName="SANS2D")
```

File: Framework/LiveData/test/KafkaHistoStreamDecoderTest.h [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+
+  void test_Histo_Stream() {
+    using namespace ::testing;
+    using namespace KafkaTesting;
+    using Mantid::API::Workspace_sptr;
+    using Mantid::DataObjects::Workspace2D;
+    using namespace Mantid::LiveData;
+
+    auto mockBroker = std::make_shared<MockKafkaBroker>();
+    EXPECT_CALL(*mockBroker, subscribe_(_, _))
```
Comment (OwenArnold) [inline Framework/LiveData/test/KafkaHistoStreamDecoderTest.h]: Not clear to my why this is being called 3 times or why different return types are provided on each call - why is this necessary for the test?

File: Framework/LiveData/inc/MantidLiveData/Kafka/IKafkaStreamDecoder.h [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+  void startCapture(bool startNow = true);
+  void stopCapture() noexcept;
+  ///@}
+
+  ///@name Querying
+  ///@{
+  bool isCapturing() const noexcept { return m_capturing; }
+  virtual bool hasData() const noexcept = 0;
+  int runNumber() const noexcept { return m_runNumber; }
+  virtual bool hasReachedEndOfRun() noexcept = 0;
```
Comment (MikeHart85) [inline Framework/LiveData/inc/MantidLiveData/Kafka/IKafkaStreamDecoder.h]: My understanding of a function declaring itself `noexcept` is that it means:

> I promise not to allow any exceptions to escape my scope, and if I break that promise then **please call std::terminate** for me

This is a runtime feature, not a compile-time check like `const`. Is that really the intent here and the other places it is used?

The only time I'm aware of that this has a significant functional benefit is when marking constructors `noexcept`, which allows libraries like the STL to optimize (re)allocations using things like [move_if_noexcept()](https://en.cppreference.com/w/cpp/utility/move_if_noexcept) to avoid copies.

Other notable properties and side-effects:

- Functions marked `noexcept` can still call functions that aren't `noexcept`, and use exceptions internally
- Only difference is that failing to `catch` one and allowing it to propagate results in `std::terminate`
- Once a virtual function is declared `noexcept`, derived classes cannot override this
- Destructors are always `noexcept` by default, unless explicitly marked `noexcept(false)`

It looks like this in the old KafkaEventStreamDecoder, so not sure this is something that needs to / should be addressed in this PR. Still, something worth considering I think.

**Review by MikeHart85 — COMMENTED** [review]
As mentioned, not sure this PR is the best place to address this, so I'm marking this a comment rather than a change request.

Couldn't see any other problems. Tested as described above and seems to work as intended.

File: Framework/LiveData/test/KafkaEventStreamDecoderTest.h [hunk 14]
```diff
@@ -66,9 +66,8 @@ class KafkaEventStreamDecoderTest : public CxxTest::TestSuite {
     using namespace Mantid::LiveData;
 
     auto mockBroker = std::make_shared<MockKafkaBroker>();
-    Mock::VerifyAndClearExpectations(&mockBroker);
     EXPECT_CALL(*mockBroker, subscribe_(_, _))
-        .Times(Exactly(3))
+        .Times(Exactly(2))
```
Comment (OwenArnold) [inline Framework/LiveData/test/KafkaEventStreamDecoderTest.h]: Looks good, but despite being restricted to 2 calls. It is set up to return 3 outputs.

Comment (LamarMoore) [reply Framework/LiveData/test/KafkaEventStreamDecoderTest.h]: That was is mistake on my part. I was testing to make sure it would fail. Need to change this back.

Comment (OwenArnold) [reply Framework/LiveData/test/KafkaEventStreamDecoderTest.h]: 👍

**Review by OwenArnold — DISMISSED** [review]

**Comment by LamarMoore** [comment]
@OwenArnold I have fixed the unit tests. Think the `KafkaEventStreamDecoderTest` verification should be handled in another PR.

**Review by OwenArnold — APPROVED** [review]
